ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/jca/XATerminatorUnitTest.java

```diff
@@ -374,4 +383,67 @@ public void testInvalid () throws Exception
         {
         }
     }
+
+    private AtomicInteger completionCount;
+
+    @Before
+    public void setup() {
+        completionCount = new AtomicInteger(0);
+    }
+
+    @Test
+    public void testConcurrentImport () throws Exception {
+        XidImple xid = new XidImple(new Uid());
+
+        final int TASK_COUNT = 400;
+        final int THREAD_COUNT = 200;
+        final CyclicBarrier gate = new CyclicBarrier(THREAD_COUNT + 1);
+
+        ArrayList<CompletableFuture<SubordinateTransaction>> futures = new ArrayList<>();
+        ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
+
+        for (int i = 0; i < TASK_COUNT; i++)
+            futures.add(doAsync(gate, i < THREAD_COUNT, executor, xid));
+
+        gate.await();
+
+        SubordinateTransaction prevStx = null;
+
+        for (CompletableFuture<SubordinateTransaction> future : futures) {
+            SubordinateTransaction stx = future.get();
+            if (stx == null) {
+                fail("transaction import returned null for future ");
+            } else {
+                if (prevStx != null)
+                    assertEquals("transaction import for same xid returned a different instance", stx, prevStx);
+                else
+                    prevStx = stx;
+            }
+        }
+
+        assertEquals("some transaction import futures did not complete", completionCount.get(), TASK_COUNT);
+    }
+
+    /*
+     * import a transaction asynchronously to maximise the opportunity for concurrency errors in TransactionImporterImple
+     */
+    private CompletableFuture<SubordinateTransaction> doAsync(
+            final CyclicBarrier gate, final boolean wait, ExecutorService executor, final XidImple xid) {
+        return CompletableFuture.supplyAsync(new Supplier<SubordinateTransaction>() {
+            @Override
+            public SubordinateTransaction get() {
+                try {
+                    if (wait)
+                        gate.await();
+                    SubordinateTransaction stx = SubordinationManager.getTransactionImporter().importTransaction(xid);
```

Comment:
I wonder if some byteman rule atEntry on _transactions.putIfAbsent would be useful? That said then we are unit testing the putIfAbsent code.